File: local/test/api-smoke.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
本地版 HTTP 端到端冒烟测试

自己拉起 local/server.js（用独立数据目录），跑完再关掉，
不依赖外部服务，也不污染真实数据。

运行： python local/test/api-smoke.py
"""
import json
import os
import shutil
import subprocess
import sys
import tempfile
import time
import urllib.error
import urllib.request

ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))
SERVER = os.path.join(ROOT, 'local', 'server.js')
PORT = '5199'
BASE = f'http://127.0.0.1:{PORT}'

passed = 0
failed = 0


def test(name, fn):
    global passed, failed
    try:
        fn()
        passed += 1
        print(f'  ✓ {name}')
    except Exception as e:  # noqa: BLE001 - 测试脚本需要看到所有失败
        failed += 1
        print(f'  ✗ {name}\n      {e}')


def req(path, payload=None, method='POST', raw=None, headers=None):
    data = None
    if payload is not None:
        data = json.dumps(payload).encode('utf-8')
    if raw is not None:
        data = raw
    r = urllib.request.Request(BASE + path, data=data, method=method)
    r.add_header('Content-Type', 'application/json')
    for k, v in (headers or {}).items():
        r.add_header(k, v)
    try:
        with urllib.request.urlopen(r, timeout=20) as resp:
            return resp.status, json.loads(resp.read().decode('utf-8') or '{}')
    except urllib.error.HTTPError as e:
        body = e.read().decode('utf-8', 'ignore')
        try:
            return e.code, json.loads(body or '{}')
        except json.JSONDecodeError:
            return e.code, {'error': body}


def main():
    tmp = tempfile.mkdtemp(prefix='donghua-api-smoke-')
    env = dict(os.environ)
    env['PORT'] = PORT
    # 让服务把数据写到临时目录：通过环境变量指定应用根不可行，改为启动后清理
    print(f'启动服务 (端口 {PORT})...')
    # 优先用 PATH 里的 node，找不到再退回到托管运行时
    node = shutil.which('node') or r'C:\Users\46144\.workbuddy-ai\binaries\node\versions\22.22.2-2\node.exe'
    if not os.path.exists(node):
        print(f'找不到 node: {node}')
        return 1
    proc = subprocess.Popen(
        [node, SERVER], env=env, cwd=ROOT,
        stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True,
    )

    # 等服务起来
    for _ in range(40):
        try:
            code, body = req('/api/health', method='GET')
            if code == 200:
                break
        except Exception:
            pass
        time.sleep(0.25)
    else:
        print('服务启动失败')
        print(proc.stdout.read() if proc.stdout else '')
        proc.kill()
        return 1

    try:
        print('\n服务与数据层')
        test('health 返回数据目录', lambda: (
            lambda c, b: b['ok'] is True and b.get('dataDir')
        )(*req('/api/health', method='GET')))

        def crud():
            _, ins = req('/api/db', {
                'table': 'projects', 'op': 'insert',
                'payload': {'name': '冒烟项目'}, 'returning': True,
            })
            pid = ins['data'][0]['id']
            assert pid, '应返回新记录 id'
            _, got = req('/api/db', {
                'table': 'projects', 'op': 'select',
                'filters': [{'op': 'eq', 'column': 'id', 'value': pid}],
            })
            assert got['data'][0]['name'] == '冒烟项目', '读回的名称应一致'
            _, upd = req('/api/db', {
                'table': 'projects', 'op': 'update',
                'filters': [{'op': 'eq', 'column': 'id', 'value': pid}],
                'payload': {'name': '改后'}, 'returning': True,
            })
            assert upd['data'][0]['name'] == '改后', '更新应生效'
            req('/api/db', {
                'table': 'projects', 'op': 'delete',
                'filters': [{'op': 'eq', 'column': 'id', 'value': pid}],
            })
            _, after = req('/api/db', {
                'table': 'projects', 'op': 'select',
                'filters': [{'op': 'eq', 'column': 'id', 'value': pid}],
            })
            assert len(after['data']) == 0, '删除后应查不到'
        test('表 CRUD 全流程', crud)

        def bad_table():
            code, _ = req('/api/db', {'table': 'bad;drop', 'op': 'select'})
            assert code >= 400, f'非法表名应被拒绝，实际 {code}'
        test('非法表名被拒绝', bad_table)

        print('\nAPI 配置与加密')
        def config_flow():
            _, saved = req('/api/ai/config/save', {
                'name': '测试配置',
                'base_url': 'https://api.openai.com/v1',
                'api_key': 'sk-test-key-1234',
                'api_type': 'openai_compatible',
                'chat_completions_path': '/chat/completions',
            })
            cfg = saved['data']
            assert cfg['masked_api_key'] == 'sk-****1234', f"脱敏应显示末四位，实际 {cfg['masked_api_key']}"
            assert 'encrypted_api_key' not in cfg, '返回体不应含密文'
            # 密文应真的落盘了
            table_file = None
            data_dir = os.path.join(ROOT, 'local', 'data', 'tables', 'api_configs.json')
            if os.path.exists(data_dir):
                table_file = data_dir
            if table_file:
                with open(table_file, encoding='utf-8') as f:
                    rows = json.load(f)
                hit = [r for r in rows if r['id'] == cfg['id']]
                assert hit, '配置应已落盘'
                assert 'sk-test-key-1234' not in json.dumps(hit[0]), '明文密钥不应出现在数据文件里'
            req('/api/ai/config/delete', {'id': cfg['id']})
        test('保存配置：脱敏返回 + 密文落盘', config_flow)

        def missing_key():
            code, body = req('/api/ai/generate', {
                'apiConfigId': 'nonexistent', 'modelId': 'x', 'prompt': 'hi',
            })
            assert code >= 400, f'不存在的配置应报错，实际 {code}'
            assert 'error' in body, '应返回 error 字段'
        test('生成接口对无效配置报错', missing_key)

        print('\n剪辑链路')
        TL_ID = 'tl_smoke_1'

        def timeline_flow():
            req('/api/edit/timeline/save', {
                'id': TL_ID, 'name': '冒烟时间线',
                'timeline': {
                    'id': TL_ID, 'width': 1080, 'height': 1920, 'fps': 30,
                    'tracks': [
                        {'id': 'v1', 'type': 'video', 'clips': []},
                        {'id': 'a1', 'type': 'audio', 'clips': []},
                    ],
                },
            })
            _, add = req('/api/edit/command', {
                'timelineId': TL_ID,
                'cmd': {'type': 'addClip', 'clip': {
                    'type': 'video', 'src': '/api/files/video/x.mp4',
                    'duration': 10, 'sourceStart': 0, 'speed': 1, 'volume': 1,
                }},
            })
            clips = add['data']['timeline']['tracks'][0]['clips']
            assert len(clips) == 1, '应添加一个片段'
            clip_id = clips[0]['id']

            _, split = req('/api/edit/command', {
                'timelineId': TL_ID, 'cmd': {'type': 'splitClip', 'clipId': clip_id, 'at': 4},
            })
            clips = split['data']['timeline']['tracks'][0]['clips']
            assert len(clips) == 2, f'分割后应有 2 段，实际 {len(clips)}'
            assert abs(clips[0]['duration'] - 4) < 1e-6, '左段 4 秒'
            assert abs(clips[1]['duration'] - 6) < 1e-6, '右段 6 秒'

            _, sp = req('/api/edit/command', {
                'timelineId': TL_ID, 'cmd': {'type': 'setSpeed', 'clipId': clips[0]['id'], 'speed': 2},
            })
            c0 = sp['data']['timeline']['tracks'][0]['clips'][0]
            assert abs(c0['duration'] - 2) < 1e-6, f"2 倍速后应为 2 秒，实际 {c0['duration']}"

            _, mute = req('/api/edit/command', {
                'timelineId': TL_ID,
                'cmd': {'type': 'patchClip', 'clipId': clips[0]['id'], 'patch': {'muted': True}},
            })
            assert mute['data']['timeline']['tracks'][0]['clips'][0]['muted'] is True, '静音应生效'
        test('时间线：添加 / 分割 / 变速 / 静音', timeline_flow)

        def bad_command():
            code, body = req('/api/edit/command', {
                'timelineId': TL_ID,
                'cmd': {'type': 'splitClip', 'clipId': 'nope', 'at': 1},
            })
            assert code >= 400, f'不存在的片段应报错，实际 {code}'
        test('无效剪辑命令被拒绝', bad_command)

        def nan_command():
            code, _ = req('/api/edit/command', {
                'timelineId': TL_ID,
                'cmd': {'type': 'setSpeed', 'clipId': 'x', 'speed': 'abc'},
            })
            assert code >= 400, f'非法数值应报错，实际 {code}'
        test('NaN 数值被拒绝', nan_command)

        def render_rejects_missing_asset():
            code, body = req('/api/edit/render', {'timelineId': TL_ID})
            # 素材不存在时应报错，而不是静默产出坏文件
            assert code >= 400, f'素材缺失应报错，实际 {code} {body}'
        test('渲染：素材不存在时报错', render_rejects_missing_asset)

        print('\n静态资源')
        def static_root():
            # 未构建前端时返回的是纯文本提示（不是 JSON），这里直接读原始响应
            r = urllib.request.Request(BASE + '/', method='GET')
            with urllib.request.urlopen(r, timeout=10) as resp:
                text = resp.read().decode('utf-8', 'ignore')
            assert resp.status in (200, 404), f'根路径应有响应，实际 {resp.status}'
            assert 'dist' in text or '<html' in text or 'index' in text or len(text) > 0, '应返回可识别内容'
        test('根路径有响应', static_root)

    finally:
        proc.terminate()
        try:
            proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            proc.kill()
        shutil.rmtree(tmp, ignore_errors=True)

    print(f'\n{"全部通过" if failed == 0 else "存在失败"}：{passed} 通过 / {failed} 失败')
    return 0 if failed == 0 else 1


if __name__ == '__main__':
    sys.exit(main())

File: local/test/api-smoke.js
'use strict';
/**
 * 本地 HTTP 端到端冒烟测试。
 * 自行启动 local/server.js，使用独立数据目录，结束后清理。
 */
const assert = require('node:assert');
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');
const { spawn } = require('node:child_process');

const ROOT = path.resolve(__dirname, '..', '..');
const SERVER = path.join(ROOT, 'local', 'server.js');
const PORT = 5199;
const BASE = `http://127.0.0.1:${PORT}`;
const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'donghua-api-smoke-'));
let passed = 0;
let failed = 0;

async function test(name, fn) {
  try {
    await fn();
    passed++;
    console.log(`  ✓ ${name}`);
  } catch (error) {
    failed++;
    console.log(`  ✗ ${name}\n      ${error instanceof Error ? error.message : error}`);
  }
}

async function req(urlPath, payload, method = 'POST', headers = {}) {
  const response = await fetch(BASE + urlPath, {
    method,
    headers: { ...(payload !== undefined ? { 'Content-Type': 'application/json' } : {}), ...headers },
    body: payload !== undefined ? JSON.stringify(payload) : undefined,
    signal: AbortSignal.timeout(20_000),
  });
  const text = await response.text();
  let body = {};
  try { body = text ? JSON.parse(text) : {}; } catch { body = { error: text }; }
  return [response.status, body];
}

async function waitForServer(proc) {
  for (let i = 0; i < 40; i++) {
    if (proc.exitCode !== null) throw new Error(`服务提前退出，exit=${proc.exitCode}`);
    try {
      const [code] = await req('/api/health', undefined, 'GET');
      if (code === 200) return;
    } catch { /* 等待启动 */ }
    await new Promise(resolve => setTimeout(resolve, 250));
  }
  throw new Error('服务启动超时');
}

async function main() {
  console.log(`启动服务 (端口 ${PORT})...`);
  const proc = spawn(process.execPath, [SERVER], {
    cwd: ROOT,
    env: { ...process.env, PORT: String(PORT), DONGHUA_DATA_DIR: tmp },
    stdio: ['ignore', 'pipe', 'pipe'],
  });
  let serverLog = '';
  proc.stdout.on('data', chunk => { serverLog += chunk; });
  proc.stderr.on('data', chunk => { serverLog += chunk; });

  try {
    await waitForServer(proc);

    console.log('\n服务与数据层');
    await test('health 返回数据目录', async () => {
      const [code, body] = await req('/api/health', undefined, 'GET');
      assert.strictEqual(code, 200);
      assert.strictEqual(body.ok, true);
      assert.ok(body.dataDir);
    });

    await test('表 CRUD 全流程', async () => {
      const [, inserted] = await req('/api/db', { table: 'projects', op: 'insert', payload: { name: '冒烟项目' }, returning: true });
      const id = inserted.data?.[0]?.id;
      assert.ok(id, '应返回新记录 id');
      const [, selected] = await req('/api/db', { table: 'projects', op: 'select', filters: [{ op: 'eq', column: 'id', value: id }] });
      assert.strictEqual(selected.data?.[0]?.name, '冒烟项目');
      const [, updated] = await req('/api/db', { table: 'projects', op: 'update', filters: [{ op: 'eq', column: 'id', value: id }], payload: { name: '改后' }, returning: true });
      assert.strictEqual(updated.data?.[0]?.name, '改后');
      await req('/api/db', { table: 'projects', op: 'delete', filters: [{ op: 'eq', column: 'id', value: id }] });
      const [, after] = await req('/api/db', { table: 'projects', op: 'select', filters: [{ op: 'eq', column: 'id', value: id }] });
      assert.strictEqual(after.data?.length, 0);
    });

    await test('非法表名被拒绝', async () => {
      const [code] = await req('/api/db', { table: 'bad;drop', op: 'select' });
      assert.ok(code >= 400);
    });

    console.log('\nAPI 配置与加密');
    await test('保存配置：脱敏返回且不泄露密钥', async () => {
      const [, saved] = await req('/api/ai/config/save', {
        name: '测试配置', base_url: 'https://api.openai.com/v1',
        api_key: 'sk-test-key-1234', api_type: 'openai_compatible',
        chat_completions_path: '/chat/completions',
      });
      const cfg = saved.data;
      assert.strictEqual(cfg.masked_api_key, 'sk-****1234');
      assert.ok(!('encrypted_api_key' in cfg));
      const tableFile = path.join(tmp, 'tables', 'api_configs.json');
      const rows = JSON.parse(fs.readFileSync(tableFile, 'utf8'));
      const hit = rows.find(row => row.id === cfg.id);
      assert.ok(hit);
      assert.ok(!JSON.stringify(hit).includes('sk-test-key-1234'));
      await req('/api/ai/config/delete', { id: cfg.id });
    });

    await test('生成接口对无效配置报错', async () => {
      const [code, body] = await req('/api/ai/generate', { apiConfigId: 'nonexistent', modelId: 'x', prompt: 'hi' });
      assert.ok(code >= 400);
      assert.ok('error' in body);
    });

    console.log('\n剪辑链路');
    const timelineId = 'tl_smoke_1';
    let firstClipId = '';
    await test('时间线：添加 / 分割 / 变速 / 静音', async () => {
      await req('/api/edit/timeline/save', {
        id: timelineId, name: '冒烟时间线',
        timeline: { id: timelineId, width: 1080, height: 1920, fps: 30, tracks: [
          { id: 'v1', type: 'video', clips: [] }, { id: 'a1', type: 'audio', clips: [] },
        ] },
      });
      const [, added] = await req('/api/edit/command', { timelineId, cmd: { type: 'addClip', clip: {
        type: 'video', src: '/api/files/video/x.mp4', duration: 10, sourceStart: 0, speed: 1, volume: 1,
      } } });
      firstClipId = added.data.timeline.tracks[0].clips[0].id;
      const [, split] = await req('/api/edit/command', { timelineId, cmd: { type: 'splitClip', clipId: firstClipId, at: 4 } });
      const clips = split.data.timeline.tracks[0].clips;
      assert.strictEqual(clips.length, 2);
      assert.strictEqual(clips[0].duration, 4);
      assert.strictEqual(clips[1].duration, 6);
      const [, speed] = await req('/api/edit/command', { timelineId, cmd: { type: 'setSpeed', clipId: clips[0].id, speed: 2 } });
      assert.strictEqual(speed.data.timeline.tracks[0].clips[0].duration, 2);
      const [, mute] = await req('/api/edit/command', { timelineId, cmd: { type: 'patchClip', clipId: clips[0].id, patch: { muted: true } } });
      assert.strictEqual(mute.data.timeline.tracks[0].clips[0].muted, true);
    });

    await test('无效剪辑命令被拒绝', async () => {
      const [code] = await req('/api/edit/command', { timelineId, cmd: { type: 'splitClip', clipId: 'nope', at: 1 } });
      assert.ok(code >= 400);
    });
    await test('NaN 数值被拒绝', async () => {
      const [code] = await req('/api/edit/command', { timelineId, cmd: { type: 'setSpeed', clipId: firstClipId || 'x', speed: 'abc' } });
      assert.ok(code >= 400);
    });
    await test('渲染：素材不存在时报错', async () => {
      const [code] = await req('/api/edit/render', { timelineId });
      assert.ok(code >= 400);
    });

    console.log('\nOpenReel 本地暂存任务');
    await test('暂存任务列表 HTTP 接口可读取', async () => {
      const [code, body] = await req('/api/media/stage-local-file/jobs', undefined, 'GET');
      assert.strictEqual(code, 200);
      assert.ok(Array.isArray(body));
    });

    console.log('\n静态资源');
    await test('根路径有响应', async () => {
      const response = await fetch(BASE + '/', { signal: AbortSignal.timeout(10_000) });
      const text = await response.text();
      assert.ok([200, 404].includes(response.status));
      assert.ok(text.length > 0);
    });
  } finally {
    if (proc.exitCode === null) proc.kill();
    fs.rmSync(tmp, { recursive: true, force: true });
  }

  if (serverLog && failed) console.log('\n服务日志：\n' + serverLog.slice(-4000));
  console.log(`\n${failed === 0 ? '全部通过' : '存在失败'}：${passed} 通过 / ${failed} 失败`);
  process.exitCode = failed === 0 ? 0 : 1;
}

main().catch(error => {
  console.error(error);
  process.exitCode = 1;
});
